authn.go

```diff
@@ -71,58 +70,30 @@ func Errorf(template string, args ...any) *connect.Error {
 	return connect.NewError(connect.CodeUnauthenticated, fmt.Errorf(template, args...))
 }
 
-// Request describes a single RPC invocation.
-type Request struct {
-	request *http.Request
-}
-
-// BasicAuth returns the username and password provided in the request's
-// Authorization header, if any.
-func (r Request) BasicAuth() (username string, password string, ok bool) {
-	return r.request.BasicAuth()
-}
-
-// Cookies parses and returns the HTTP cookies sent with the request, if any.
-func (r Request) Cookies() []*http.Cookie {
-	return r.request.Cookies()
-}
-
-// Cookie returns the named cookie provided in the request or
-// [http.ErrNoCookie] if not found. If multiple cookies match the given name,
-// only one cookie will be returned.
-func (r Request) Cookie(name string) (*http.Cookie, error) {
-	return r.request.Cookie(name)
-}
-
 // Procedure returns the RPC procedure name, in the form "/service/method". If
 // the request path does not contain a procedure name, the entire path is
 // returned.
-func (r Request) Procedure() string {
-	path := strings.TrimSuffix(r.request.URL.Path, "/")
+func Procedure(request *http.Request) string {
+	path := strings.TrimSuffix(request.URL.Path, "/")
 	ultimate := strings.LastIndex(path, "/")
 	if ultimate < 0 {
-		return r.request.URL.Path
+		return request.URL.Path
 	}
 	penultimate := strings.LastIndex(path[:ultimate], "/")
 	if penultimate < 0 {
-		return r.request.URL.Path
+		return request.URL.Path
 	}
 	procedure := path[penultimate:]
 	if len(procedure) < 4 { // two slashes + service + method
-		return r.request.URL.Path
+		return request.URL.Path
 	}
 	return procedure
 }
 
-// ClientAddr returns the client address, in IP:port format.
-func (r Request) ClientAddr() string {
-	return r.request.RemoteAddr
-}
-
 // Protocol returns the RPC protocol. It is one of [connect.ProtocolConnect],
 // [connect.ProtocolGRPC], or [connect.ProtocolGRPCWeb].
-func (r Request) Protocol() string {
-	ct := r.request.Header.Get("Content-Type")
+func Protocol(request *http.Request) string {
```

Comment:
I would prefer to add this function to the Connect runtime, since we seem to need it in quite a few places. In Connect, we should choose a name for the function that clarifies that it's inferring/guessing the procedure - maybe something like `InferProcedure` or `NewProcedureFromURL`. We should probably have the function also take just the URL rather than the whole request.

Reply:
Have created a PR to add two methods, one for procedure and the other for protocol: https://github.com/connectrpc/connect-go/pull/756